File: src/main/java/org/example/gomoku/GameBoard.java
package org.example.gomoku;

/**
 * 五子棋棋盘逻辑
 */
public class GameBoard {

    public static final int SIZE = 15;
    public static final int EMPTY = 0;
    public static final int BLACK = 1;
    public static final int WHITE = 2;

    private final int[][] board;
    private int currentPlayer;
    private boolean gameOver;
    private int winner;
    private int[][] winLine;

    public GameBoard() {
        board = new int[SIZE][SIZE];
        reset();
    }

    public void reset() {
        for (int[] row : board) {
            java.util.Arrays.fill(row, EMPTY);
        }
        currentPlayer = BLACK;
        gameOver = false;
        winner = EMPTY;
        winLine = null;
    }

    /**
     * 在指定位置落子
     * @return true 如果落子成功
     */
    public boolean placeStone(int row, int col) {
        if (gameOver || row < 0 || row >= SIZE || col < 0 || col >= SIZE) {
            return false;
        }
        if (board[row][col] != EMPTY) {
            return false;
        }
        board[row][col] = currentPlayer;

        // 检查是否获胜
        if (checkWin(row, col)) {
            winner = currentPlayer;
            gameOver = true;
        } else {
            currentPlayer = (currentPlayer == BLACK) ? WHITE : BLACK;
        }
        return true;
    }

    private boolean checkWin(int row, int col) {
        int[][] directions = {{0, 1}, {1, 0}, {1, 1}, {1, -1}};
        int player = board[row][col];

        for (int[] dir : directions) {
            int dr = dir[0], dc = dir[1];
            int count = 1;
            int startRow = row, startCol = col;
            int endRow = row, endCol = col;

            // 正方向
            for (int i = 1; i < 5; i++) {
                int r = row + dr * i, c = col + dc * i;
                if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] == player) {
                    count++;
                    endRow = r;
                    endCol = c;
                } else {
                    break;
                }
            }
            // 反方向
            for (int i = 1; i < 5; i++) {
                int r = row - dr * i, c = col - dc * i;
                if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] == player) {
                    count++;
                    startRow = r;
                    startCol = c;
                } else {
                    break;
                }
            }

            if (count >= 5) {
                winLine = new int[][]{{startRow, startCol}, {endRow, endCol}};
                return true;
            }
        }
        return false;
    }

    public int getCell(int row, int col) {
        return board[row][col];
    }

    public int getCurrentPlayer() {
        return currentPlayer;
    }

    public boolean isGameOver() {
        return gameOver;
    }

    public int getWinner() {
        return winner;
    }

    public int[][] getWinLine() {
        return winLine;
    }
}

File: ios/Gomoku/GameBoard.swift
import Foundation
import Combine

class GameBoard: ObservableObject {
    static let size = 15
    static let empty = 0
    static let black = 1
    static let white = 2

    @Published var board: [[Int]]
    @Published var currentPlayer: Int
    @Published var gameOver: Bool
    @Published var winner: Int
    @Published var winLine: [(row: Int, col: Int)]
    @Published var moveHistory: [(row: Int, col: Int, player: Int)]

    init() {
        board = Array(repeating: Array(repeating: 0, count: GameBoard.size), count: GameBoard.size)
        currentPlayer = GameBoard.black
        gameOver = false
        winner = 0
        winLine = []
        moveHistory = []
    }

    func reset() {
        board = Array(repeating: Array(repeating: 0, count: GameBoard.size), count: GameBoard.size)
        currentPlayer = GameBoard.black
        gameOver = false
        winner = 0
        winLine = []
        moveHistory = []
    }

    func placeStone(row: Int, col: Int) -> Bool {
        guard row >= 0, row < GameBoard.size,
              col >= 0, col < GameBoard.size,
              board[row][col] == GameBoard.empty,
              !gameOver else {
            return false
        }

        board[row][col] = currentPlayer
        moveHistory.append((row: row, col: col, player: currentPlayer))

        if checkWin(row: row, col: col) {
            gameOver = true
            winner = currentPlayer
        } else if moveHistory.count == GameBoard.size * GameBoard.size {
            gameOver = true
            winner = 0
        } else {
            currentPlayer = currentPlayer == GameBoard.black ? GameBoard.white : GameBoard.black
        }

        return true
    }

    func checkWin(row: Int, col: Int) -> Bool {
        let player = board[row][col]
        let directions = [(0, 1), (1, 0), (1, 1), (1, -1)]

        for (dr, dc) in directions {
            var line: [(row: Int, col: Int)] = [(row: row, col: col)]

            // Check forward
            for i in 1...4 {
                let r = row + dr * i
                let c = col + dc * i
                guard r >= 0, r < GameBoard.size,
                      c >= 0, c < GameBoard.size,
                      board[r][c] == player else { break }
                line.append((row: r, col: c))
            }

            // Check backward
            for i in 1...4 {
                let r = row - dr * i
                let c = col - dc * i
                guard r >= 0, r < GameBoard.size,
                      c >= 0, c < GameBoard.size,
                      board[r][c] == player else { break }
                line.append((row: r, col: c))
            }

            if line.count >= 5 {
                winLine = line
                return true
            }
        }

        winLine = []
        return false
    }

    func undoLastMove() -> Bool {
        guard let lastMove = moveHistory.popLast() else {
            return false
        }

        board[lastMove.row][lastMove.col] = GameBoard.empty
        currentPlayer = lastMove.player
        gameOver = false
        winner = 0
        winLine = []
        return true
    }

    func getBoardCopy() -> [[Int]] {
        return board.map { $0 }
    }
}
